lightbeam/validate.py

```diff
@@ -18,19 +33,190 @@ def __init__(self, lightbeam=None):
+        references_structure = self.rebalance_local_references_structure(references_structure)
+        # more memory-efficient to load local data and populate cache for one endpoint at a time:
+        for original_endpoint in references_structure.keys():
+            endpoints_to_check = self.EDFI_GENERICS_TO_RESOURCES_MAPPING.get(original_endpoint, [original_endpoint])
+            for endpoint in endpoints_to_check:
+                if endpoint in self.local_reference_cache.keys():
+                    # already loaded (when validating another endpoint); no need to reload
+                    continue
+                self.logger.debug(f"(discovering any local data for {endpoint}...)")
+                endpoint_data = self.load_references_data(endpoint, references_structure)
+                self.local_reference_cache[endpoint] = self.references_data_to_cache(endpoint, endpoint_data, references_structure)
+
+    # this is (unfortunately) necessary to allow lookup of nested references in local payload
+    # (for remote reference lookup, a flat dict of keys is passed to the Ed-Fi API and it takes care of nesting)
+    def rebalance_local_references_structure(self, references_structure):
+        if "objectiveAssessments" in references_structure.keys():
+            if "assessmentIdentifier" in references_structure["objectiveAssessments"]:
+                references_structure["objectiveAssessments"].remove("assessmentIdentifier")
+                references_structure["objectiveAssessments"].append("assessmentReference.assessmentIdentifier")
+            if "namespace" in references_structure["objectiveAssessments"]:
+                references_structure["objectiveAssessments"].remove("namespace")
+                references_structure["objectiveAssessments"].append("assessmentReference.namespace")
+        return references_structure
+
+    def references_data_to_cache(self, endpoint, endpoint_data, references_structure):
+        cache = []
+        structure = references_structure[endpoint]
+        for payload in endpoint_data:
+            sorted_keys = structure.copy()
+            sorted_keys.sort(key=lambda x: x.split(".")[-1])
+            cache_key = ''
+            for key in sorted_keys:
+                cache_key += f"{payload[key]}~~~"
+            cache.append(cache_key)
+        return cache
+
+    def load_references_data(self, endpoint, references_structure):
+        data = []
+        data_files = self.lightbeam.get_data_files_for_endpoint(endpoint)
+        for file_name in data_files:
+            with open(file_name) as file:
+                for line_counter, line in enumerate(file):
+                    line = line.strip()
+                    try:
+                        payload = json.loads(line)
+                    except Exception as e:
+                        self.logger.warning(f"... (ignoring invalid JSON payload at {line_counter} of {file_name})")
```

Comment:
Depends on how `lightbeam validate` is called. If with selectors and the referenced resource is not included, then yes it would be ignored. But if no selectors (all data in `data_dir` to be validated) or the referenced resource is also selected, then when the referenced resource is validated, lightbeam would throw validation errors for such a malformed payload.

Maybe an example to illuminate this: if you have the following in your `data_dir`:
* `schools.jsonl` with payloads that reference a LocalEducationAgency
* `localEducationAgencies.jsonl` with a payload that's not valid JSON

Then
* `lightbeam validate -s schools` would just log `(ignoring invalid JSON payload at X of localEducationAgencies.jsonl)`
* `lightbeam validate -s schools,localEducationAgencies` or simply `lightbeam validate` would log `VALIDATION ERROR (line X): invalid JSON`

I guess I think this behavior is reasonable. What do you think @johncmerfeld ?

Reply:
Ah, gotcha. I missed that nuance; makes sense to me!